pkg/api/service/warnings.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"fmt"
+	"net/netip"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/apis/core/helper"
+)
+
+func GetWarningsForService(service, oldService *api.Service) []string {
+	if service == nil {
+		return nil
+	}
+	var warnings []string
+
+	if helper.IsServiceIPSet(service) {
+		warnings = append(warnings, getWarningsForClusterIPs(service)...)
+	}
+
+	if service.Spec.Type == api.ServiceTypeExternalName {
+		warnings = append(warnings, getWarningsForExternalName(service)...)
+	}
+
+	if oldService != nil {
+		warnings = append(warnings, getWarningsForServiceTransitions(service, oldService)...)
+	}
+
+	return warnings
+}
+
+// getWarningsForClusterIPs returns warnings for ClusterIPs
+func getWarningsForClusterIPs(service *api.Service) []string {
+	var warnings []string
+	var fieldPath *field.Path
+
+	for i, clusterIP := range service.Spec.ClusterIPs {
+		// IPs with leading zeros CVE-2021-29923 are not valid in golang since 1.17
+		// This will also warn about possible future changes on the golang std library
+		// xref: https://issues.k8s.io/108074
+		ip, err := netip.ParseAddr(clusterIP)
+		if err != nil {
+			warnings = append(warnings, fmt.Sprintf("%s: https://issues.k8s.io/108074, IP address %s invalid in a future release: %v", fieldPath.Child("spec", "clusterIPs").Index(i), clusterIP, err))
+			continue
+		}
+		// A Recommendation for IPv6 Address Text Representation
+		//
+		// "All of the above examples represent the same IPv6 address.  This
+		// flexibility has caused many problems for operators, systems
+		// engineers, and customers.
+		// ..."
+		// https://datatracker.ietf.org/doc/rfc5952/
+		if ip.Is6() && ip.String() != clusterIP {
+			warnings = append(warnings, fmt.Sprintf("%s: IP address %s not in canonical format %s (RFC 5952), invalid in a future release", fieldPath.Child("spec", "clusterIPs").Index(i), clusterIP, ip.String()))
```

Comment:
RFC 5952 says "all implementations MUST accept and be able to handle any legitimate RFC4291 format". We can perhaps warn if they use a bad format, but I don't think we should reject them ever.

Reply:
(I'm happy not to warn and just allow the API to canonify.)

Reply:
we can not canonify, it increase the risk of hotlooping on controllers (we had that problem)

1. controllers sets service.Spec.ClusterIP = 2001:db8:0:0:0:0::1
2. apiserver process it and stores service.Spec.ClusterIP = 2001:db8::1
3. controller reconcile and reads the service 2001:db8::1 != 2001:db8:0:0:0:0::1
4. is not my IP, back to point 1

We can say, you have to parse and compare IPs, but reality is different and a lot of implementations compare strings

Reply:
> but I don't think we should reject them ever.

I just added some drama 😄 , but I don't have strong opinion, open to suggestions
- warn about the risk of hotlooping
- warn just to do a recommendation
- no warn at all
- ..

Reply:
warn, but we'll never make it invalid

Reply:
I suppose we could report the non-canonicalness in `.status`, but doubt that we care enough to make an API change.